### ISSUE TITLE
HMLANGW mode doesn't work in 3.69.7.20230506

### ISSUE THREAD

**Issue by not1337** [jens-maus/RaspberryMatic#2340]
### Describe the issue you are experiencing

/bin/hmlangw hangs in an endless communication loop over /dev/mmd_bidcos at startup, thus no lan gateway functionality (checked with strace). Downgrade to 3.67.10.20230114 fixes the issue.

### Describe the behavior you expected

lan gateway mode

### Steps to reproduce the issue

Fresh install of 3.69.7.20230506 on a Pi 3B with HM-RF-MOD, configure networking, ntp server and ssh, create /usr/local/HMLGW, reboot.

### What is the version this bug report is based on?

3.69.7.20230506

### Which base platform are you running?

rpi3 (RaspberryPi3)

### Which HomeMatic/homematicIP radio module are you using?

RPI-RF-MOD

### Anything in the logs that might be useful for us?

```txt
none as /bin/hmlangw just loops at 100% cpu
```

### Additional information

I just ordered another Pi 3B / HM-RF-MOD combo on which I could do some testing before this system goes live as a lan gateway, if you are interested. Should arrive in a couple of days.

**Comment by jens-maus** [comment]
Sorry, but this can't be a general / common issue since several other people are constantly using the LAN Gateway mode of RaspberryMatic with a `HM-MOD-RPI-PCB` or with a `RPI-RF-MOD` and the last time I checked it also worked for me. So please investigate further to rule out local issues on your side.

**Comment by Baxxy13** [comment]
No Problems here with LGW-Mode on Pi2B with HM-MOD-RPI-PCB.
(running Nightly 3.69.7.20230612)

**Comment by not1337** [comment]
Sorry for closing, was not intendend.

**Comment by jens-maus** [comment]
Please explain what a `HM-RF-MOD` is? I guess you are meaning the `RPI-RF-MOD` but as you always call it `HM-RF-MOD` I want to make sure which exact RF module you are indeed using as I never heard of any `HM-RF-MOD`...

**Comment by not1337** [comment]
Yes, RPI-RF-MOD, I do mess with this name constantly as everything else starts with HM...

**Comment by jens-maus** [comment]
Ok, but then simply keep on using 3.67.10.20230114 for the HMLANGW version of your RaspberryPi as there is no benefit in updating to any later version. This `HMLANGW` mode is a kind of internal "gimmick" to run RaspberryMatic in a strong stripped down mode as a dumb LAN-Gateway and nothing more. And since the `hmlangw` source code was never modified/touched since its initial release in 2015 I don't see any reason to update to newer versions if you just run RaspberryMatic in HMLANGW mode. In addition, using a RaspberryPi3+RPI-RF-MOD just for HMLANGW mode is IMHO a waste of resources. A RaspberryPi1 or even a RaspberryPi2 with a `HM-MOD-RPI-PCB` is completely enough for the LAN Gateway mode. Thus, keep this version running forever if it works for you and perhaps someone will find the time to try to reproduce your issue. But as @Baxxy13 already stated here at least with the `HM-MOD-RPI-PCB` he does not see any issue. However, perhaps he can try to reproduce the same with a `RPI-RF-MOD` and then report here.

**Comment by not1337** [comment]
No problem here, it should however be stated as a hint in the documentation that for RPI-RF-MOD the working image is 3.67.10. BTW I'm using this hardware to have a replacement around in case the Raspberrymatic system dies due to a hardware fault.

**Comment by jens-maus** [comment]
> No problem here, it should however be stated as a hint in the documentation that for RPI-RF-MOD the working image is 3.67.10.

This hasn't been verified yet by another person, thus no need to hurry for any documentation adaptions. Wait for someone to reproduce it here and then we will see...

**Comment by Baxxy13** [comment]
Sorry Jens. "Debugging" `S47InitRFHardware` ist a bit too much for me.

**Comment by jens-maus** [comment]
Ok, I tested it myself now with a RaspberryPi2 and a `RPI-RF-MOD` and here it seems to work:

```
Identifying Homematic RF-Hardware: ....HmRF: RPI-RF-MOD/GPIO@3f201000.serial, HmIP: RPI-RF-MOD/GPIO@3f201000.serial, OK
Updating Homematic RF-Hardware: RPI-RF-MOD: 4.4.22, not necessary, OK
```

```
# netstat -tulpn | grep LISTEN
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1393/sshd [listener
tcp        0      0 0.0.0.0:2000            0.0.0.0:*               LISTEN      1480/hmlangw
tcp        0      0 0.0.0.0:2001            0.0.0.0:*               LISTEN      1480/hmlangw
tcp        0      0 :::22                   :::*                    LISTEN      1393/sshd [listener
```

Thus, `hmlangw` seem to start up properly and is listening for connections on port 2000/2001. However, lets see if it works with a RaspberryPi3b as well.

**Comment by jens-maus** [comment]
After I have tested the LAN-Gateway mode myself yesterday (including a RaspberryPi3+RPI-RF-MOD environment) I can at least partly support the observations by @not1337. In fact, sometimes it seems the `hmlangw` cannot startup properly and is endlessly waiting for input data on `/dev/mmd_bidcos` during startup before it usually goes on in processing and then starts up additional threads to listen on port 2000 and 2001. See here:

```
# /bin/hmlangw -n 58A992F000 -s /dev/mmd_bidcos -r -1 -D
serial fd 3 name /dev/mmd_bidcos
```

Usually it should continue outputting info on thread startup and showing receiving bytes entering via `/dev/mmd_bidcos`. And I could also reproduce that also sometimes it correctly starts up even on a Pi3+RPI-RF-MOD combo.

However, the issue does not seem to be 100% reproducible and sometimes it requires multiple restarts of `hmlangw` before it is not able to receive any more input data from `/dev/mmd_bidcos` and is waiting forevery during the startup phase. What can be seen, however, is that during startup `hmlangw` seem to get flooded by events on the eq3loop char devices as the kernel regularly spits out `not enough space in buffers` error/warning messages which, however, could be simply a sympton of the startup problem:

```
[ 1893.707889] eq3loop: eq3loop_write_master() mmd_bidcos: not enough space in buffers. free space = 10, required space = 18
[ 1893.707904] eq3loop: eq3loop_write_master() return error: -14
```

Perhaps this also partly points at some synchronization issue in the eq3loop char kernel driver?!?!

Long story short: There really seem to be a startup/synchronization issue when hmlangw tries to startup and this really needs further investigation. And perhaps it would be nice if @Baxxy13 or @jp112sdl could try to reproduce these startup issues... However, I also see the complications with all this `hmlangw` functionality and that it still has low priority, but if we have some luck this could be a simple startup/synchronization issue in the hmlangw source code which we perhaps can easily solve in the end...

**Comment by mhoellein** [comment]
Here is also a problem with the LAN-Gateway. I have four LAN-Gateway connected via WLAN an LAN. Two LAN-Gateways are original Homematic HM-LGW-O-TW-W-EU and the other LAN-Gateways are Rasperry Pis where I run the binary hmlangw. This works for many years now. If the LAN-Gateway disconnects, they can´t be reconnect to the raspberrian version 3.69.7.20230506. Downgrade to Version 3.69.6.20230407 fix the problem for me. On version 3.69.7.20230506 many times the Webinterface shows connected but the DutyCycle shows -1 and no device is via the LAN-Gateway reachable.

**Comment by jens-maus** [comment]
With the next nightly snapshot a new version of the `hmlangw` LAN gateway daemon will be included in which I fixed an issue during bootloader setup/init which should explain the issues @not1337 reported here. So please try to use the next nightly snapshot and see if it improves the `hmlangw` startup/workings.

**Comment by bronkoo** [comment]
Hi,

just updated my LAN gateway (`HM-MOD-RPI-PCB` on `RaspberryPi Zero W`) from 3.69.7.20230506 to 3.69.7.20230626 and it stops working exactly by the issue described here for these previous version, where I'm personally have had no problems, but in this version never checked log messages because it worked.

ports are open:
```
# netstat -tulpn | grep LISTEN
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1245/sshd [listener
tcp        2      0 0.0.0.0:2000            0.0.0.0:*               LISTEN      1332/hmlangw
tcp        0      0 0.0.0.0:2001            0.0.0.0:*               LISTEN      1332/hmlangw
tcp        0      0 :::22                   :::*                    LISTEN      1245/sshd [listener

# tail -f /var/log/hmlangw.log 
Client 192.168.178.33 connected to BidCos port!
Client 192.168.178.33 connected to keepalive port!
Keepalive client closed connection.
BidCos client closed connection.
``` 

but system hangs on crowded /var/log/messages:

```
Jun 28 11:11:27 homematic-raspi-hgw kern.err kernel: [ 4333.530009] eq3loop: eq3loop_write_master() mmd_bidcos: not enough space in buffers. free space = 15, required space = 18
Jun 28 11:11:27 homematic-raspi-hgw kern.info kernel: [ 4333.530076] eq3loop: eq3loop_write_master() return error: -14
``` 
How can I help for further investigations?

**Comment by jens-maus** [comment]
Sorry, but this issue is closed here. If you have a new problem, open a new issue and explain your issue in more detail. From your posting I am not sure if you actually do have a real problem or not or if you are just interpreting the kernel messages as a kind of error but you do not have a practical error using the hmlangw. These kernel messages (`.. not enough space in buffers...`) are not automatically point at a problem but are just symptoms of data flooding that could easily happen during initial startup and synchronization.... 

So the more important point is if the LAN gateway connection works or not, and if it does and your device are reachable and working in long term: fine. So please report issues based on real problem and not by trying to simply interpret log file entries as problems.... This is unfortunately a common mistake these days...

**Comment by jens-maus** [comment]
If it worked with previous versions and now with the new version it does not, yes. Open a new issue, but please show detail technical information on the previous case (where it worked) and now with the new version where it stopped working and please try to identify more details on the actual behaviour (100% cpu use?, no successful connection in the main raspberrymatic webui?, etc. etc.)

**Comment by not1337** [comment]
Sorry for being late, just got around testing. The good news is that with the 20230630 nightly the issue is solved.

**Comment by jens-maus** [comment]
> Sorry for being late, just got around testing. The good news is that with the 20230630 nightly the issue is solved.

This does not make sense. From 3.69.7.20230626 to 3.69.7.20230630 nothing HMLANGW relevant was changed. These versions are essentially the same.

**Comment by not1337** [comment]
I simply couldn't get around to test an earlier version than 20230630 - time constraints.

**Comment by bronkoo** [comment]
Sorry, no test on 20230630 here because of production system (care system for my old parents ;-)...
